src/StreamJsonRpc/Reflection/MessageFormatterProgressTracker.cs

```diff
@@ -0,0 +1,239 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace StreamJsonRpc.Reflection
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Reflection;
+    using Microsoft;
+    using Microsoft.VisualStudio.Threading;
+
+    /// <summary>
+    /// Class containing useful methods to help message formatters implement support for <see cref="IProgress{T}"/>.
+    /// </summary>
+    public class MessageFormatterProgressTracker
+    {
+        /// <summary>
+        /// Special method name for progress notification.
+        /// </summary>
+        public const string ProgressRequestSpecialMethod = "$/progress";
+
+        /// <summary>
+        /// Dictionary to record the calculation made in <see cref="FindIProgressOfT"/> to obtain the <see cref="IProgress{T}"/> type from a given <see cref="Type"/>.
+        /// </summary>
+        private static readonly Dictionary<Type, Type> IProgressOfTTypeMap = new Dictionary<Type, Type>();
+
+        /// <summary>
+        /// Dictionary used to map the outbound request id to their progress id token so that the progress objects are cleaned after getting the final response.
+        /// </summary>
+        private readonly Dictionary<long, object> requestProgressMap = new Dictionary<long, object>();
+
+        /// <summary>
+        /// Dictionary used to map progress id token to its corresponding <see cref="ProgressParamInformation" /> instance containing the progress object and the necessary fields to report the results.
+        /// </summary>
+        private readonly Dictionary<object, ProgressParamInformation> progressMap = new Dictionary<object, ProgressParamInformation>();
+
+        /// <summary>
+        /// Object used to lock the access to <see cref="requestProgressMap"/> and <see cref="progressMap"/>.
+        /// </summary>
+        private readonly object progressLock = new object();
+
+        /// <summary>
+        /// Gets or Sets the id of the request currently being serialized so the converter can use it to create the request-progress map.
+        /// </summary>
+        public long? RequestIdBeingSerialized { get; set; }
+
+        /// <summary>
+        /// Gets or sets the the next id value to assign as token for the progress objects.
+        /// </summary>
+        private long NextProgressId { get; set; }
+
+        /// <summary>
+        /// Converts given <see cref="Type"/> to its <see cref="IProgress{T}"/> type.
+        /// </summary>
+        /// <param name="objectType">The type which may implement <see cref="IProgress{T}"/>.</param>
+        /// <returns>The <see cref="IProgress{T}"/> from given <see cref="Type"/> object, or <c>null</c>  if no such interface was found in the given <paramref name="objectType" />.</returns>
+        public static Type FindIProgressOfT(Type objectType)
```

Comment:
>FindIProgressOfT [](start = 27, length = 16)

The only caller of this in a formatter appears to use it to test whether the result is non-null, and they want to know whether `AddProgressObjectToMap` will accept the object later. So how about making this method private (to support your own uses), and adding a new public method called `IsSupportedProgressType` that returns `bool` so that callers know better when to call it and have less code to write?